internal/analysis/individual.go

```diff
@@ -20,220 +19,76 @@ import (
-		}
-	}
+	job := jobmanager.Job[string, *model.IndividualAnalysis]{
+		JobOptions:              &options.JobOptions,
+		LockKey:                 fmt.Sprintf("analysis-individual-course-%s", lockCourseID),
+		PoolSize:                config.ANALYSIS_INDIVIDUAL_COURSE_POOL_SIZE.Get(),
+		ReturnIncompleteResults: !options.WaitForCompletion,
+		WorkItems:               fullSubmissionIDs,
+		RetrieveFunc:            db.GetIndividualAnalysis,
+		StoreFunc:               db.StoreIndividualAnalysis,
+		RemoveFunc:              db.RemoveIndividualAnalysis,
+		WorkFunc: func(fullSubmissionID string) (*model.IndividualAnalysis, error) {
+			return computeSingleIndividualAnalysis(options, fullSubmissionID, true)
+		},
+		WorkItemKeyFunc: func(fullSubmissionID string) string {
+			return fmt.Sprintf("analysis-individual-%s", fullSubmissionID)
+		},
+		OnComplete: func(result *jobmanager.JobOutput[string, *model.IndividualAnalysis]) {
+			if result == nil {
+				return
+			}
 
-	poolSize := config.ANALYSIS_INDIVIDUAL_COURSE_POOL_SIZE.Get()
-	type AnalysisPoolResult struct {
-		Result  *model.IndividualAnalysis
-		RunTime int64
+			collectIndividualStats(fullSubmissionIDs, result.RunTime, options.InitiatorEmail)
+		},
 	}
 
-	poolResults, err := util.RunParallelPoolMap(poolSize, fullSubmissionIDs, options.Context, func(fullSubmissionID string) (AnalysisPoolResult, error) {
-		result, runTime, err := runSingleIndividualAnalysis(options, fullSubmissionID)
-		if err != nil {
-			err = fmt.Errorf("Failed to perform individual analysis on submission %s: '%w'.", fullSubmissionID, err)
-		}
-
-		return AnalysisPoolResult{result, runTime}, err
-	})
-
+	err = job.Validate()
 	if err != nil {
-		return nil, fmt.Errorf("Failed to run analysis in a parallel pool: '%w'.", err)
+		return nil, 0, fmt.Errorf("Failed to validate job: '%w'.", err)
 	}
 
-	// If the analysis was canceled, exit right away.
-	if poolResults.Canceled {
-		return nil, nil
+	output := job.Run()
+	if output.Error != nil {
+		return nil, 0, fmt.Errorf("Failed to run individual analysis job '%s': '%w'.", output.ID, output.Error)
 	}
 
-	poolResults.IsDone()
-
-	var errs error = nil
-	totalRunTime := int64(0)
-
-	for _, id := range fullSubmissionIDs {
-		err, ok := poolResults.WorkErrors[id]
-		if ok {
-			errs = errors.Join(errs, err)
-			continue
-		}
-
-		analysisPoolResult, ok := poolResults.Results[id]
-		if !ok {
-			errs = errors.Join(errs, fmt.Errorf("Unable to find result for submission: '%s'.", id))
-			continue
-		}
+	if len(output.WorkErrors) != 0 {
+		for fullSubmissionID, err := range output.WorkErrors {
+			courseID, assignmentID, email, shortSubmissionID, splitErr := common.SplitFullSubmissionID(fullSubmissionID)
+			if splitErr != nil {
+				log.Error("Failed to split full submission ID.", splitErr, log.NewAttr("submission-id", fullSubmissionID))
+				log.Error("Failed to run individual analysis.", err, log.NewAttr("submission-id", fullSubmissionID), log.NewAttr("job-id", output.ID))
 
```

Comment:
This looks a tad messy (not because of you, just the nature of the situation).

Would it be possible to make a function that returns `[]*log.Attr` and then splat/unpack (`...`) them into the log statement?